libraries/chain/chain_controller.cpp

```diff
@@ -872,51 +878,84 @@ void chain_controller::validate_expiration(const Transaction& trx) const
 } FC_CAPTURE_AND_RETHROW((trx)) }
 
 uint32_t chain_controller::_transaction_message_rate(uint32_t now, uint32_t last_update_sec, uint32_t rate_limit_time_frame_sec,
-                                                     uint32_t rate_limit, uint32_t previous_rate, const char* type, const AccountName& name)
+                                                     uint32_t rate_limit, uint32_t previous_rate, rate_limit_type type, const AccountName& name)
 {
    const auto delta_time = now - last_update_sec;
    uint32_t message_count = 1;
    if (delta_time <= rate_limit_time_frame_sec)
    {
       message_count += ( ( ( rate_limit_time_frame_sec - delta_time ) * fc::uint128( previous_rate ) )
                      / rate_limit_time_frame_sec ).to_uint64();
-      EOS_ASSERT(message_count <= rate_limit, tx_msgs_exceeded,
-                 "Rate limiting ${type}=${name} messages sent, ${count} exceeds ${max} messages limit per ${sec} seconds. Wait 1 second and try again",
-                 ("type",type)
-                 ("name",name)
-                 ("count",message_count)
-                 ("max",rate_limit)
-                 ("sec", rate_limit_time_frame_sec));
+#define RATE_LIMIT_ASSERT(tx_msgs_exceeded, type_str) \
+      EOS_ASSERT(message_count <= rate_limit, tx_msgs_exceeded, \
+                 "Rate limiting ${type} account=${name} messages sent, ${count} exceeds ${max} messages limit per ${sec} seconds. Wait 1 second and try again", \
+                 ("type",type_str) \
+                 ("name",name) \
+                 ("count",message_count) \
+                 ("max",rate_limit) \
+                 ("sec", rate_limit_time_frame_sec))
+      switch (type)
+      {
+      case authorization_account:
+         RATE_LIMIT_ASSERT(tx_msgs_auth_exceeded, "authorization");
+         break;
+      case code_account:
+         RATE_LIMIT_ASSERT(tx_msgs_code_exceeded, "code");
+         break;
+      }
    }
 
    return message_count;
 }
 
 void chain_controller::rate_limit_message(const Message& message)
 { try {
+   const auto now = head_block_time().sec_since_epoch();
+
+   // per authorization rate limiting
    for (const auto& permission : message.authorization)
    {
       auto rate_limiting = _db.find<rate_limiting_object, by_name>(permission.account);
-      const auto now = head_block_time().sec_since_epoch();
       if (rate_limiting == nullptr)
       {
          _db.create<rate_limiting_object>([&](rate_limiting_object& rlo) {
             rlo.name = permission.account;
-            rlo.trans_msg_rate_per_account = 1;
-            rlo.last_update_sec = now;
+            rlo.per_auth_account_trans_msg_rate = 1;
+            rlo.per_auth_account_last_update_sec = now;
```

Comment:
Make per_auth_account_last_update_sec be a time_point_sec instead of uint32_t. Also for types in chain_controller, basically everywhere use time_point_sec instead of raw uint32_t.